### ISSUE TITLE
BUG: new reference passed to non-stealing API (static analyzer report)

### ISSUE THREAD

**Issue by Snape3058** [scipy/scipy#16232]
1. New reference returned and assigned to `obj`.
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L46

2. Assume `obj` is not NULL.
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L47

3. `obj` passed to [`PyDict_SetItemString`](https://docs.python.org/3/c-api/dict.html#c.PyDict_SetItemString), which will not steal a reference from `obj`.
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L53

4. Function returns without decreasing the refcnt of `obj`.
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L55

Internal Report ID: fd73a1

**Comment by rgommers** [comment]
On closer inspection, this may be a bug but the analysis of step 4 is a little off.

> 4. Function returns without decreasing the refcnt of `obj`.

The most common pattern for PyDict_SetItemString is:
```C
    r = PyDict_SetItemString(thread_dict, key, (PyObject *)obj);
    Py_DECREF(obj)
    return r;   /* r is 0 or 1 here */
```

However, note that in our case, `obj` itself is returned:
```C
    PyDict_SetItemString(thread_dict, key, (PyObject *)obj);
    return obj;
```
Decreasing the refcount of an object before returning it is not necessary, returning usually transfers ownership. If we'd decref it here, its refcount would be 0 and it could be garbage collected straight away.

The callers of `get_tls_global` are not decref'ing the returned `obj` either however. So that may be the actual problem here? This code is a little tricky to understand, so I'm not 100% sure. @Snape3058 does this sound right to you?

**Comment by Snape3058** [comment]
Sorry for presenting a partial program path.

Line 46 returns a new reference and is assigned to `obj`. (event 1)
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L46
Then, the reference is not stolen on line 53 with the call to `PyDict_SetItemString`. (event 3)
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L53
So the return on line 55 will return the new reference (acquired on line 46), rather than a borrowed reference like line 43, to `g` on line 145. (event 4)
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L55
However, it is not decreased in the function with variable `g`.
https://github.com/scipy/scipy/blob/8627df31ab07d66b438c7e5d5880e3f4d435f248/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L145

For function `get_tls_global`, it is also confusing as it can return both a borrowed reference (line 43) and a new reference (line 55).
And for its caller, when a borrowed reference is returned, the code is correct. (as you mentioned in the previous posts)
But when a new reference is returned as mentioned above, the reference will get leaked.

**Comment by rgommers** [comment]
> However, it is not decreased in the function with variable `g`.

It shouldn't be decreased there - then we'd again get to refcount 0 and risk the dict entry getting cleaned up while it is still needed.

> But when a new reference is returned as mentioned above, the reference will get leaked.

I think the point here is not that the reference gets leaked. What that actually means is "the reference count does not reach zero, and hence the memory it points to is not freed by Python's garbage collection mechanism". In this case though, the `thread_dict` owns the memory of its elements and every element is cleaned up when the thread shuts down in `superlu_python_module_free`.

In other words, leaking the reference is effectively making `PyDict_SetItemString` steal ownership.

> For function `get_tls_global`, it is also confusing as it can return both a borrowed reference (line 43) and a new reference (line 55).

I agree it's confusing. There's probably no way static analysis is going to understand this code construct. But it does seem correct to me (and hard to change).

**Comment by Snape3058** [comment]
> > However, it is not decreased in the function with variable `g`.
> 
> It shouldn't be decreased there - then we'd again get to refcount 0 and risk the dict entry getting cleaned up while it is still needed.

Yes, you are right.
I do not mean decrease the refcnt via `g`, but `obj`.

> > But when a new reference is returned as mentioned above, the reference will get leaked.
> 
> I think the point here is not that the reference gets leaked. What that actually means is "the reference count does not reach zero, and hence the memory it points to is not freed by Python's garbage collection mechanism". In this case though, the `thread_dict` owns the memory of its elements and every element is cleaned up when the thread shuts down in `superlu_python_module_free`.
> 
> In other words, leaking the reference is effectively making `PyDict_SetItemString` steal ownership.

`PyDict_SetItemString` **DOES NOT STEAL** a reference to the object inserted.
It means the API will increase the refcnt first, and it is not needed to preserve a reference for the dict.
And the destruction of the dict will only release the reference the dict holds, rather than the reference you preserved.

---

Let's see the following example.

```c
#define PY_SSIZE_T_CLEAN
#include <Python.h>

#define printrefcnt(op)                                                     \
  do {                                                                      \
    PyObject *__p = op;                                                     \
    printf("%s:%d refcnt(%s:0x%lx) = %ld\n", __PRETTY_FUNCTION__, __LINE__, \
           #op, (uintptr_t)__p, Py_REFCNT(__p));                            \
  } while (0)

PyObject *foo() {
  PyObject *d = PyDict_New();
  PyObject *x = PyLong_FromLong(5239);
  printrefcnt(x);  // > foo:14 refcnt(x:0x7fc2e54f0330) = 1
  PyDict_SetItemString(d, "key-x", x);
  printrefcnt(x);  // > foo:16 refcnt(x:0x7fc2e54f0330) = 2
  Py_DECREF(x);
  printrefcnt(x);  // > foo:18 refcnt(x:0x7fc2e54f0330) = 1
  return x;  // Returning a borrowed reference from dict `d`
}
```

The output is noted in the comments.
As you can see, after the PyObject `x` is created, its refcnt is 1. (`foo` is holding the reference)
Then it is inserted into the dict, and the refcnt is increased to 2 in the API call. (one held by `foo`, one held by `d`)
Then the refcnt decrement releases the reference held by `foo` and the refcnt is changed to 1. (`d` is now holding the reference)
And the final return will return a borrowed reference from dict `d`.

Just the same as the case I reported.
`x` is `obj`, without the decrement in my example, the returned x will be a new reference held in `foo`.
As it is not proper to decrease the refcnt via `g`, it should be decreased in `get_tls_global` like my example.
It can make the returned object also a borrowed reference like other branches.

I know decreasing the refcnt in `superlu_python_module_free` via `g` is not correct.
You can add a decrement after line 53 to make the returned PyObject also a borrowed reference like other branches.
And to make it also safe when the insertion fails, you can do this.

```c
if (PyDict_SetItemString(d, key, x)) return NULL;
Py_DECREF(x);
return x;
```

What do you think about my suggestions?

**Comment by czgdp1807** [comment]
Actually, I agree with @Snape3058. I did my own analysis and found that refcount of `obj` is 2. So there is a memory leak indeed (or reference leak). Why? Because when the `thread_dict` will be destroyed then the refcount of `obj` will reduce to 1 (from 2). So there will be a leak. We should do `Py_DECREF(obj)` to ensure that there is only one reference of `obj` i.e., inside the `thread_dict`.

I applied the following diff to SciPy,

```diff
diff --git a/scipy/sparse/linalg/_dsolve/_superlu_utils.c b/scipy/sparse/linalg/_dsolve/_superlu_utils.c
index 49b928a43..a2bb82511 100644
--- a/scipy/sparse/linalg/_dsolve/_superlu_utils.c
+++ b/scipy/sparse/linalg/_dsolve/_superlu_utils.c
@@ -27,6 +27,16 @@ static SuperLUGlobalObject *get_tls_global(void)
     }
     return &superlu_py_global;
 #else
+
+    #define printrefcnt(op)                                                     \
+        do {                                                                     \
+            PyObject *__p = (PyObject*) op;                                                   \
+            printf("scipy/sparse/linalg/_dsolve/_superlu_utils.c:%d %s:%d refcnt(%s:0x%lx) = %ld\n", __LINE__, \
+                __PRETTY_FUNCTION__, __LINE__, \
+                #op, (uintptr_t)__p, Py_REFCNT(__p));                            \
+        } while (0);
+
+
     PyObject *thread_dict;
     SuperLUGlobalObject *obj;
     const char *key = "scipy.sparse.linalg._dsolve._superlu.__global_object";
@@ -40,10 +50,12 @@ static SuperLUGlobalObject *get_tls_global(void)
 
     obj = (SuperLUGlobalObject*)PyDict_GetItemString(thread_dict, key);
     if (obj && Py_TYPE(obj) == &SuperLUGlobalType) {
+        printrefcnt(obj)
         return obj;
     }
 
     obj = (SuperLUGlobalObject*)PyObject_New(SuperLUGlobalObject, &SuperLUGlobalType);
+    printrefcnt(obj)
     if (obj == NULL) {
         return (SuperLUGlobalObject*)PyErr_NoMemory();
     }
@@ -51,7 +63,7 @@ static SuperLUGlobalObject *get_tls_global(void)
     obj->jmpbuf_valid = 0;
 
     PyDict_SetItemString(thread_dict, key, (PyObject *)obj);
-
+    printrefcnt(obj)
     return obj;
 #endif
 }
```

then I executed the following code,

```python
import sys

from numpy import eye
import numpy.random as random
from numpy.testing import assert_equal

from scipy.sparse import csc_array
from scipy.sparse.linalg._dsolve import splu

# Test that we are keeping track of the reference count with splu.
n = 30
a = random.random((n, n))
a[a < 0.95] = 0
# Make a diagonal dominant, to make sure it is not singular
a += 4*eye(n)
a_ = csc_array(a)
lu = splu(a_)

# And now test that we don't have a refcount bug
rc = sys.getrefcount(lu)
for attr in ('perm_r', 'perm_c'):
    perm = getattr(lu, attr)
    assert_equal(sys.getrefcount(lu), rc + 1)
    del perm
    assert_equal(sys.getrefcount(lu), rc)
```

and got the following output,

<Details>

<Summary>Output log</Summary>

```zsh
(scipy-dev-latest) 18:25:52:~/Quansight/scipy % python dev.py python ../debug.py
💻  ninja -C /Users/czgdp1807/Quansight/scipy/build -j10
ninja: Entering directory `/Users/czgdp1807/Quansight/scipy/build'
[4/4] Linking target scipy/sparse/linalg/_dsolve/_superlu.cpython-313-darwin.so
Build OK
💻  meson install -C build --only-changed --tags runtime,python-runtime,tests,devel
Installing, see meson-install.log...
Installation OK
scipy/sparse/linalg/_dsolve/_superlu_utils.c:58 SuperLUGlobalObject *get_tls_global(void):58 refcnt(obj:0x11df44ff0) = 1
scipy/sparse/linalg/_dsolve/_superlu_utils.c:66 SuperLUGlobalObject *get_tls_global(void):66 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
scipy/sparse/linalg/_dsolve/_superlu_utils.c:53 SuperLUGlobalObject *get_tls_global(void):53 refcnt(obj:0x11df44ff0) = 2
```

</Details>


So you can see, the reference count is `2`. Also, at,

https://github.com/scipy/scipy/blob/9ec1da20e44c13109e4ce04120ff9f642eb03518/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L46

the reference count of `obj` is 1. At,

https://github.com/scipy/scipy/blob/9ec1da20e44c13109e4ce04120ff9f642eb03518/scipy/sparse/linalg/_dsolve/_superlu_utils.c#L53

the reference count of `obj` is 2. `PyDict_SetItemString` increases the reference count of `obj`. Therefore if we `Py_DECREF(obj)` before doing `return obj` then reference count of `obj` will become `1` (this one reference belongs to the `thread_dict` I think). When `thread_dict` will be destroyed then refcount of `obj` will become 0 and it will be garbage collected. So everything looks safe to me.

cc: @Snape3058 @rgommers

**Comment by rgommers** [comment]
Sorry, I completely lost track of this issue. It sounds like I was wrong and @Snape3058 was 100% correct. A PR with a fix would be great.

**Comment by czgdp1807** [comment]
Sure. Will send a PR tomorrow.
